rfcs/text/0428-extending-types.md

```diff
@@ -0,0 +1,677 @@
+      name◄┐              │
+           ├───Person─────Employee (subtype)
+       age◄┘              │
+                          │
+occupation◄───────────────┘
+```
+
+**An example of _compatible_, overlapping properties**:
+
+- Supertype `Person` contains the required properties `Name` and `Age`
+- Supertype `Superhero` contains the required properties `Superpower` and `Name`
+
+In this example, `Name` overlaps as a required property in both supertypes.
+
+```txt
+              (supertypes)
+superpower◄────Superhero──┐
+               │          │
+      name◄────┤          │
+               │          │
+       age◄────Person─────Employee (subtype)
+                          │
+occupation◄───────────────┘
+
+
+```
+
+**An example of _incompatible_, overlapping properties**:
+
+- Supertype `Person` contains the required properties `Name` and `Age`
+- Supertype `Superhero` contains the required property `Superpower` and an array of `Name`s
+
+In this example, the array of `Name`s on the `Superhero` type would not be compatible with the required `Name` property of `Person`, which means that the two types cannot be supertypes together.
+
+## Additional properties on types
+
+In the proposed [Versioning RFC](./0408-versioning-types.md) for the type system, having `{ "additionalProperties": false }` for all schemas is an assumption made for determining type compatibility, which means that any supertype will not validate against a subtype that adds properties if it receives all properties of a subtype instance. For example, if we supply the `Employee` instance from above to a `Person`, it will receive properties that are considered `additionalProperties` (the `Occupation` property is not present on `Person`).
+
+The assumption that we can select/project parts of a subtype that make up a supertype is essential for keeping strictness in JSON Schemas.
+
+And unfortunately, specifying `{ "unevaluatedProperties": false }` does not behave as expected when composing types together in JSON Schema either, which means we will have to redefine how `{ "additionalProperties": true }` or `{ "unevaluatedProperties": false }` behaves within our type extension system to make supertypes keep strictness while allowing composition.
+
+Concrete examples of how JSON Schema breaks with these validation constraints are shown in the [Reference-level explanation](#additional-properties-on-types-1)
+
+## Defining extended types
+
+Extended types will be defined with conventional JSON Schema syntax, the `allOf` keyword. When creating a new entity type it's possible to extend another entity type by adding an entry to `allOf` value with a versioned URI reference.
+Using a versioned URI makes it so that subtypes aren't automatically updated when the supertype is.
+
+As extended types can extend other extended types, we must also make sure that there are no cycles within the type hierarchy, as it could lead to hard to reason about types and unpredictability.
```

Comment:
This was a realization I had after writing this section initially, and you're right. There isn't an actual issue with cycles because of how versioning is defined, and the cycles would always end up with a "leaf type" that isn't circular if we resolve deeply enough. 

I still dislike the idea of having circular extension, though, but yet to find a very good reason to not allow it